Runtime/XRMeshChain.cs

```diff
@@ -0,0 +1,349 @@
+            /// Refresh positions
+            /// </summary>
+            Positions = 1,
+
+            /// <summary>
+            /// Refresh colors
+            /// </summary>
+            Colors = 2,
+
+            /// <summary>
+            /// Refresh sizes
+            /// </summary>
+            Sizes = 4,
+
+            /// <summary>
+            /// Refresh all mesh components
+            /// </summary>
+            All = 7
+        }
+
+        Vector3[] m_Verts;
+        Color32[] m_Colors;
+        List<Vector4> m_ShapeData; // xy: UV coordinates for GPU expansion zw: Size of this vertex, size of the neighbor
+        List<Vector3> m_NeighborPoints; // Location of the next point this pipe connects to, or itself if it is a billboard
+
+        // Update flags to prevent unnecessary mesh data generation
+        MeshRefreshFlag m_DataThatNeedsUpdate = MeshRefreshFlag.All;
+
+        // Cached runtime data
+        Mesh m_Mesh;
+        Transform m_OwnerTransform;
+
+        bool m_WorldSpaceData;
+
+        /// <summary>
+        /// Whether the control points of this mesh chain have been referenced in world or local space
+        /// This makes sure the bounding box of the mesh is updated appropriately for culling
+        /// </summary>
+        public bool worldSpaceData
+        {
+            get { return m_WorldSpaceData; }
+            set { m_WorldSpaceData = value; }
+        }
+
+        /// <summary>
+        /// How many primitives/quads this mesh chain supports and has reserved memory for
+        /// </summary>
+        public int reservedElements { get; private set; }
+
+        /// <summary>
+        /// If using world space data, this option will force the center of the bounding box to be the root
+        /// </summary>
+        public bool centerAtRoot { get; set; }
+
+        /// <summary>
+        /// Initialize a new XRMeshChain
+        /// </summary>
+        public XRMeshChain()
+        {
+            reservedElements = 0;
+        }
+
+        /// <summary>
+        /// Creates or re-creates the mesh with all the data needed for billboard-pipe based line rendering
+        /// </summary>
+        /// <param name="owner">The gameobject that will own the created mesh</param>
+        /// <param name="dynamic">Whether this mesh is going to updated frequently or not</param>
+        /// <param name="totalElements">How many total billboards and pipes are needed for this renderer</param>
+        public void GenerateMesh(GameObject owner, bool dynamic, int totalElements, bool setMesh = true)
+        {
+            // Precache neccessary data
+            // The mesh, vertex and triangle counts
+            if (m_Mesh == null)
+            {
+                m_Mesh = new Mesh();
+            }
+
+            if (dynamic == true)
+            {
+                m_Mesh.MarkDynamic();
+            }
+
+            if (setMesh)
+                owner.GetComponent<MeshFilter>().mesh = m_Mesh;
```

Comment:
This was also needed to prevent dirtying the asset in OnValidate. I'm not sure how to avoid this extra check, but ideally we should avoid doing this on the prefab instance, and only do this on instances in the scene. I'm not aware of a callback that will do this in editor and runtime the way we want.

Reply:
Actually this is probably something we can check for fairly easily.  The editor-specific can be ifdef'd out in a build, because that's a scenario where you would *never* have a prefab instance

Reply:
It's not about being in a build--packages are immutable and this dirties prefabs w/ XRLineRenderers. What this was doing is, on import, dirtying some prefabs in EXR, which was causing validation issues

Reply:
That's what I am saying - in the editor we can check if something is a prefab instance

Reply:
Oh, I see. I considered this but wasn't sure where exactly we want to check. Should we check here or earlier in `OnValidate`? Does any of this have to happen on the prefab instance, i.e. when you are first creating the prefab?